audiostream2py/data.py

```diff
@@ -0,0 +1,148 @@
+
+        Further info on methodology found in timestamping data stream discussion:
+        https://miro.com/app/board/uXjVPsuJtdM=/?share_link_id=872583858418
+
+        Note: slice 'step' is unlikely used in the traditional sense which is how it is implemented.
+        A future improvement could use 'step' as a channel selector by filtering out bytes of other
+        channels.
+
+        :param index: timestamp or slice
+        :return: Sliced AudioData
+        """
+        if isinstance(index, slice):
+            if index.start is not None:
+                start_sample, start_date = self.nearest_sample_index_and_time(
+                    timestamp=index.start, rounding_type='ceil'
+                )
+            else:
+                start_sample, start_date = 0, self.start_date
+
+            if index.stop is not None:
+                end_sample, end_date = self.nearest_sample_index_and_time(
+                    timestamp=index.stop, rounding_type='floor'
+                )
+            else:
+                end_sample, end_date = self.frame_count, self.end_date
+
+            start = start_sample * self.frame_size
+            stop = end_sample * self.frame_size
+            step = index.step * self.frame_size if index.step is not None else None
+            waveform = self.waveform[start:stop:step]
+            frame_count = end_sample - start_sample
+        else:
+            sample_idx, start_date = self.nearest_sample_index_and_time(
+                timestamp=index, rounding_type='floor'
+            )
+            _, end_date = self.nearest_sample_index_and_time(
+                timestamp=index, rounding_type='ceil'
+            )
+            waveform = self.waveform[sample_idx : sample_idx + self.frame_size]
+            frame_count = 1
+
+        return AudioData(start_date, end_date, waveform, frame_count, self.status_flags)
+
+    @cached_property
+    def frame_size(self) -> int:
+        """Byte count per frame"""
+        return int(len(self.waveform) / self.frame_count)
+
+    def nearest_sample_index_and_time(
+        self, timestamp, rounding_type='ceil'
+    ) -> Tuple[int, Union[int, float]]:
+        """Calculate sample location and round up or down to the nearest whole sample.
+
+        'ceil' for session start and 'floor' for session end
+
+        Further info on methodology found in timestamping data stream discussion:
+        https://miro.com/app/board/uXjVPsuJtdM=/?share_link_id=872583858418
+
+        :param timestamp: time of sample in the same unit as start and end date
+        :param rounding_type: 'ceil' or 'floor'
+        :return: sample_index, sample_timestamp
+        """
+        rounder = getattr(math, rounding_type)
+        sample_idx = int(
+            rounder(
+                (timestamp - self.start_date)
+                / (self.end_date - self.start_date)
+                * self.frame_count
+            )
+        ) + (0 if rounding_type == 'ceil' else 1)
```

Comment:
Not sure to understand why you're doing that. If you add 1 when`rounding_type` is "floor", it is like it was "ceil".

Reply:
Math.floor is used for slice stop which is exclusive so 1 needs to be added to include the last sample. I had to add it after failing test_process_to_find_last_sample_of_session. It was unexpected at first but it makes sense.  Also the timestamp is the end of the sample so 1 is included to account for that.  You can see in the test that I don't add 1 when calculating the expected sample value.